core/wiki/macros/tag-picker.tid

```diff
@@ -1,31 +1,50 @@
 title: $:/core/macros/tag-picker
 tags: $:/tags/Macro
 
+\define delete-tag-state-tiddlers() <$action-deletetiddler $filter="[<newTagNameTiddler>] [<storeTitle>] [<tagSelectionState>]"/>
+
 \define add-tag-actions()
-<$action-sendmessage $message="tm-add-tag" $param={{{ [<newTagNameTiddler>get[text]] }}}/>
-<$action-deletetiddler $tiddler=<<newTagNameTiddler>>/>
+<$set name="tag" value={{{ [<__tiddler__>get[text]] }}}>
+<$list filter="[<currentTiddler>!tag<tag>]" variable="ignore" emptyMessage="""
+<$action-sendmessage $message="tm-remove-tag" $param=<<tag>>/>
+""">
+<$action-sendmessage $message="tm-add-tag" $param=<<tag>>/>
+</$list>
+</$set>
+<<delete-tag-state-tiddlers>>
+<$action-setfield $tiddler=<<refreshTitle>> text="yes"/>
 \end
 
-\define tag-button()
-<$button class="tc-btn-invisible" tag="a" tooltip={{$:/language/EditTemplate/Tags/Add/Button/Hint}}>
+\define tag-button(selectedClass)
+<$button class="tc-btn-invisible $selectedClass$" tag="a" tooltip={{$:/language/EditTemplate/Tags/Add/Button/Hint}}>
 <$action-sendmessage $message="tm-add-tag" $param=<<tag>>/>
-<$action-deletetiddler $tiddler=<<newTagNameTiddler>>/>
+<<delete-tag-state-tiddlers>>
 <$macrocall $name="tag-pill" tag=<<tag>>/>
 </$button>
 \end
 
+\define clear-tags-actions()
+<$list filter="[<__storeTitle__>has[text]] [<__tiddler__>has[text]]" variable="ignore" emptyMessage="""<<delete-tag-state-tiddlers>><$action-sendmessage $message="tm-cancel-tiddler"/>""">
+<<delete-tag-state-tiddlers>>
+</$list>
+\end
+
 \define tag-picker-inner()
 \whitespace trim
+<$vars tagSelectionState=<<qualify "$:/state/selected-tag">> storeTitle=<<qualify "$:/temp/NewTagName/input">> refreshTitle=<<qualify "$:/temp/NewTagName/refresh">> nonSystemTagsFilter="[tags[]!is[system]search:title<userInput>sort[]]" systemTagsFilter="[tags[]is[system]search:title<userInput>sort[]]">
 <div class="tc-edit-add-tag">
 <span class="tc-add-tag-name">
-<$keyboard key="ENTER" actions=<<add-tag-actions>>>
-<$edit-text tiddler=<<newTagNameTiddler>> tag="input" default="" placeholder={{$:/language/EditTemplate/Tags/Add/Placeholder}} focusPopup=<<qualify "$:/state/popup/tags-auto-complete">> class="tc-edit-texteditor tc-popup-handle" tabindex=<<tabIndex>> focus={{{ [{$:/config/AutoFocus}match[tags]then[true]] ~[[false]] }}}/>
-</$keyboard>
+<$macrocall $name="keyboard-driven-input" tiddler=<<newTagNameTiddler>> tag="input" 
+		placeholder={{$:/language/EditTemplate/Tags/Add/Placeholder}} focusPopup=<<qualify "$:/state/popup/tags-auto-complete">> 
+		class="tc-edit-texteditor tc-popup-handle" tabindex=<<tabIndex>> focus={{{ [{$:/config/AutoFocus}match[tags]then[true]] ~[[false]] }}} 
+		storeTitle=<<storeTitle>> filter1=<<nonSystemTagsFilter>> filter2=<<systemTagsFilter>> inputAcceptActions=<<add-tag-actions>> 
+		inputCancelActions=<<clear-tags-actions>> filterMinLength={{$:/config/Tags/MinLength}} refreshTitle=<<refreshTitle>> 
+		selectionStateTitle=<<tagSelectionState>>/>
```

Comment:
I think it helps with readability to have all the "title" parameters first: tiddler, storeTitle, refreshTitle, selectionStateTitle